controllers/ndbserver_controller.go

```diff
@@ -37,26 +45,89 @@ type NDBServerReconciler struct {
+Reconciles the NDBServer custom resources by
+1. Checks for deletion
+2. Verify credentials and connectivity
+3. Take actions based on current status.status, fetch data
+4. Update the status if any changes are observed (excluding counter)
+*/
 func (r *NDBServerReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
-	_ = log.FromContext(ctx)
+	log := log.FromContext(ctx)
+	log.Info("NDBServer reconcile started")
+	// 1. Checks for deletion
+	// Fetch the NDBServer resource from the namespace
+	ndbServer := &ndbv1alpha1.NDBServer{}
+	err := r.Get(ctx, req.NamespacedName, ndbServer)
+	if err != nil {
+		if errors.IsNotFound(err) {
+			// Request object not found, could have been deleted after reconcile request.
+			// Return and don't requeue
+			log.Info("NDBServer resource not found. Ignoring since object must be deleted")
+			return doNotRequeue()
+		}
+		// Error reading the object - requeue the request.
+		log.Error(err, "Failed to get NDBServer")
+		return requeueOnErr(err)
+	}
+
+	// Copy the status, this will be updated by the end of the reconcile
+	status := ndbServer.Status.DeepCopy()
+	// Initialize maps if they do not exist
+	if status.Databases == nil {
+		status.Databases = make(map[string]ndbv1alpha1.NDBServerDatabaseInfo)
+	}
+
+	// log.Info("NDBServer CR Status: " + util.ToString(status))
+
+	// 2. Verify credentials and connectivity
+	// Fetch credentials and check Authentication
+	username, password, caCert, err := getNDBCredentialsFromSecret(ctx, r.Client, ndbServer.Spec.CredentialSecret, req.Namespace)
+	ndbClient := ndb_client.NewNDBClient(username, password, ndbServer.Spec.Server, caCert, ndbServer.Spec.SkipCertificateVerification)
+	if err != nil {
+		log.Error(err, "Credential Error: error while fetching credentials from CredentialSecret", "secret name", ndbServer.Spec.CredentialSecret)
+		status.Status = common.NDB_CR_STATUS_CREDENTIAL_ERROR
+	} else {
+		authResponse, err := ndb_api.AuthValidate(ctx, ndbClient)
+		if err != nil || authResponse.Status != "success" {
+			log.Error(err, "Authentication Error: Could not verify connectivity / auth credentials for NDB")
+			status.Status = common.NDB_CR_STATUS_AUTHENTICATION_ERROR
+		} else {
+			status.Status = common.NDB_CR_STATUS_OK
+		}
+	}
+
+	// 3. Take actions based on current status.status
+	switch status.Status {
+	case common.NDB_CR_STATUS_CREDENTIAL_ERROR, common.NDB_CR_STATUS_AUTHENTICATION_ERROR:
+		// no-op
+	case common.NDB_CR_STATUS_OK:
+		// Get Status (check and perform data fetching, update counters)
+		status = getNDBServerStatus(ctx, status, ndbClient)
+	default:
+		// no-op
+		return doNotRequeue()
+	}
 
-	// TODO(user): your logic here
+	// 4. Update the status if any changes are observed (excluding counter)
+	if !util.DeepEqualWithException(ndbServer.Status, *status, "Counter") {
+		log.Info("Status Changed, updating lastUpdated time")
+		status.LastUpdated = time.Now().Format(time.DateTime)
+	}
+	ndbServer.Status = *status
```

Comment:
Should this be inside the if block, since we want to update ndbServer.Status only if its changed?

Reply:
We still need to update the counter, hence this will be required on every reconcile.